contracts/minters/base-minter/src/contract.rs

```diff
@@ -107,61 +111,145 @@ pub fn execute(
     msg: ExecuteMsg,
 ) -> Result<Response, ContractError> {
     match msg {
-        ExecuteMsg::Mint { token_uri } => execute_mint_sender(deps, info, token_uri),
+        ExecuteMsg::Mint { token_uri } => execute_mint_sender(deps, env, info, token_uri),
         ExecuteMsg::UpdateStartTradingTime(time) => {
             execute_update_start_trading_time(deps, env, info, time)
         }
+        ExecuteMsg::ReceiveNft(msg) => execute_burn_to_mint(deps, env, info, msg),
     }
 }
 
+pub fn execute_burn_to_mint(
+    _deps: DepsMut,
+    env: Env,
+    info: MessageInfo,
+    msg: Cw721ReceiveMsg,
+) -> Result<Response, ContractError> {
+    let mut res = Response::new();
+    let burn_msg = cw721::Cw721ExecuteMsg::Burn {
+        token_id: msg.token_id,
+    };
+    let cosmos_burn_msg = CosmosMsg::Wasm(WasmMsg::Execute {
+        contract_addr: info.sender.to_string(),
+        msg: to_binary(&burn_msg)?,
+        funds: vec![],
+    });
+    res = res.add_message(cosmos_burn_msg);
+
+    let token_uri_msg: TokenUriMsg = from_binary(&msg.msg)?;
+    let execute_mint_msg = ExecuteMsg::Mint {
+        token_uri: token_uri_msg.token_uri,
+    };
+    let cosmos_mint_msg = CosmosMsg::Wasm(WasmMsg::Execute {
+        contract_addr: env.contract.address.to_string(),
+        msg: to_binary(&execute_mint_msg)?,
+        funds: vec![],
+    });
+    let res = res.add_message(cosmos_mint_msg);
+    Ok(res)
+}
+
 pub fn execute_mint_sender(
     deps: DepsMut,
+    env: Env,
+    info: MessageInfo,
+    token_uri: String,
+) -> Result<Response, ContractError> {
+    let config = CONFIG.load(deps.storage)?;
+
+    if matches!(config.mint_price, Token::NonFungible(_)) {
+        return Err(ContractError::InvalidMintPrice {});
+    }
+
+    _execute_mint_sender(deps, env, info, token_uri)
+}
+
+fn _check_sender_is_collection_or_contract(
+    this_contract: Addr,
+    info: MessageInfo,
+    collection_info: CollectionInfoResponse,
+) -> Result<bool, ContractError> {
+    let sender_is_collection_or_contract =
+        vec![collection_info.creator, this_contract.to_string()].contains(&info.sender.to_string());
```

Comment:
The contract itself can never be the sender. This is not a smart contract account or anything. So this can be removed and we can go back to just checking if the sender is the creator.

Reply:
I don't think this is correct.  
When run `fn execute_burn_to_mint` we first do a burn: 

```rs
    let cosmos_burn_msg = CosmosMsg::Wasm(WasmMsg::Execute {
        contract_addr: info.sender.to_string(),
        msg: to_binary(&burn_msg)?,
        funds: vec![],
    });
```
and then tack on a mint: 
```rs

    let cosmos_mint_msg = CosmosMsg::Wasm(WasmMsg::Execute {
        contract_addr: env.contract.address.to_string(),
        msg: to_binary(&execute_mint_msg)?,
        funds: vec![],
    });
    let res = res.add_message(cosmos_mint_msg);
    Ok(res)
   ```
   
   That mint message will show up as the minter executing a mint on itself and so you will end up in `pub fn execute_mint_sender` which is where we need to check if the mint message came from this minter and let it through.

Reply:
This comment is unaddressable per Discord: 

```rs
"the minters do in fact call themselves when they send a mint message to reply which ends up back in the contract and I need to check if sender is self contract"
```